packages/net-stubbing/lib/server/state.ts

```diff
@@ -1,24 +1,60 @@
-import { noop } from 'lodash'
+import type { InterceptedRequest } from './intercepted-request'
 import type { NetStubbingState } from './types'
 
 export function state (): NetStubbingState {
   return {
     requests: {},
     routes: [],
     pendingEventHandlers: {},
-    reset () {
-      // clean up requests that are still pending
-      for (const requestId in this.requests) {
-        const { res } = this.requests[requestId]
+    async reset () {
+      await Promise.all(Object.values <InterceptedRequest>(this.requests).map((request) => {
+        const responseEvents = ['before:response', 'response:callback', 'response', 'network:error']
+        const inResponsePhase = responseEvents.includes(request.lastEvent!)
+        const { res } = request
 
-        res.removeAllListeners('finish')
-        res.removeAllListeners('error')
-        res.on('error', noop)
-        res.destroy()
-      }
+        if (res.destroyed) {
+          delete this.requests[request.id]
+
+          return Promise.resolve()
+        }
+
+        let shouldDestroyResponse = false
```

Comment:
So being unfamiliar with this code, it's rather hard to follow what's going on and why. Can you comment this section a bit more. explaining what's going on and why?